File: src/eko-core/prompt/components/tools.ts
/**
 * 工具相关的提示词组件
 */

import { ComponentConfig, ComponentContext, PromptComponent } from './types'
import { resolveTemplate } from '../template-engine'

// Human Interaction component removed: handled by per-tool UI confirmation, not a standalone tool

/**
 * 工具描述组件
 */
export const toolsDescriptionComponent: ComponentConfig = {
  id: PromptComponent.TOOLS_DESCRIPTION,
  name: 'Tools Description',
  description: 'Basic description of tools',
  required: false,
  template: `# Available Tools
{toolsList}`,
  fn: async (context: ComponentContext) => {
    const { tools = [] } = context

    if (tools.length === 0) return undefined

    const toolsList = tools.map(tool => `- ${tool.name}: ${tool.description || ''}`).join('\n')

    const template = toolsDescriptionComponent.template!
    return resolveTemplate(template, { toolsList })
  },
}

// Human Interaction prompt component removed

File: src-tauri/src/agent/prompt/components/tools.rs
use async_trait::async_trait;
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;

use crate::agent::config::PromptComponent;
use crate::agent::prompt::components::types::{ComponentContext, ComponentDefinition};
use crate::agent::prompt::template_engine::TemplateEngine;
use crate::agent::{AgentError, AgentResult};

pub fn definitions() -> Vec<Arc<dyn ComponentDefinition>> {
    vec![Arc::new(ToolsDescriptionComponent)]
}

struct ToolsDescriptionComponent;

#[async_trait]
impl ComponentDefinition for ToolsDescriptionComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::ToolsDescription
    }

    fn name(&self) -> &str {
        "Tools Description"
    }

    fn description(&self) -> &str {
        "Description of available tools"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some("# Available Tools\n{tools_list}\n\nEach tool has specific parameters and usage patterns. Always provide the required parameters in the correct JSON format.")
    }

    async fn render(
        &self,
        context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        if context.tools.is_empty() {
            return Ok(None);
        }

        let template = template_override
            .or_else(|| self.default_template())
            .ok_or_else(|| AgentError::PromptBuildingError("missing template".into()))?;

        let tools_list = context
            .tools
            .iter()
            .map(|tool| format!("- {}: {}", tool.name, tool.description.clone()))
            .collect::<Vec<_>>()
            .join("\n");

        let mut template_context = HashMap::new();
        template_context.insert("tools_list".to_string(), json!(tools_list));

        let result = TemplateEngine::new()
            .resolve(template, &template_context)
            .map_err(AgentError::PromptBuildingError)?;
        Ok(Some(result))
    }
}
